core/Types.h

```diff
@@ -995,7 +995,7 @@ struct DispatchArgs {
     const TypePtr &selfType;
     const TypeAndOrigins &fullType;
     const TypePtr &thisType;
-    const std::shared_ptr<const SendAndBlockLink> &block;
+    const SendAndBlockLink *const block;
```

Comment:
There are a lot of other places where we have `shared_ptr<.*SendAndBlockLink>`. Should we change those too?

Reply:
AFAICT, all the other uses relate to threading information through the CFG, either during building, or storing it in the various CFG instructions.  We need it in the latter to share information between various instructions and we need it in the former to be able to construct the latter.

I think the use of `shared_ptr` is appropriate for this sort of thing; it's just the usage of the `const` qualifier on the contained value here that causes the interesting issues.  I suppose we could have made `DispatchArgs` carry `const std::shared_ptr<SendAndBlockLink>&`, but the `const`-qualification on the value itself seemed like valuable information.

Reply:
the second `const` in `const SendAndBlockLink *const block` just means that you can't do

```cpp
args.block = other;
```

right? And it's the first `const` that means you can't do something like

```cpp
args.block->fun = other;
```
?

Reply:
I always get these mixed up; the first one is what controls assignment to the pointer's data, while the second one is about the pointer itself: https://godbolt.org/z/77WxYxKrb

In this case having both means we can't assign to the pointer's data, and the pointer's value can't be changed after initialization, which is I think what we want.

Reply:
Yup, that's correct!  You'll sometime see people -- and I think demanglers typically do this -- write `T const &val` or `T const *val` so that, reading type declarations right-to-left, as you're supposed to, gives you "reference (resp. pointer) to `const T`".  Then `T const *const val` is a "`const` pointer to `const T`".